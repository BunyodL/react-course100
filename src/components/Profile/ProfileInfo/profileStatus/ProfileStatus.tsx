import { ProfileStatusPropsType } from './ProfileStatusWithHooks';
import React, { ChangeEvent } from 'react';
import st from './ProfileInfo.module.css';

type State = {
    editMode: boolean;
    status: string;
};

class ProfileStatus extends React.Component<ProfileStatusPropsType, State> {
    state = {
        editMode: false,
        status: this.props.status,
    };

    activateEditMode = (): void => {
        this.setState({
            editMode: true,
        });
    };

    deactivateEditMode = (): void => {
        this.setState({
            editMode: false,
        });
        this.props.updateUserStatus(this.state.status);
    };

    onStatusChange = (e: ChangeEvent<HTMLInputElement>) => {
        this.setState({
            status: e.currentTarget.value,
        });
    };

    componentDidUpdate(prevProps: ProfileStatusPropsType, prevState: State) {
        if (prevProps.status !== this.props.status) {
            this.setState({
                status: this.props.status,
            });
        }
    }

    render() {
        return (
            <div className={st.profileStatus}>
                {!this.state.editMode && (
                    <div>
                        <span onDoubleClick={this.activateEditMode}>
                            {this.props.status || '-----'}
                        </span>
                    </div>
                )}
                {this.state.editMode && (
                    <div>
                        <input
                            autoFocus
                            onChange={this.onStatusChange}
                            onBlur={this.deactivateEditMode}
                            type="text"
                            value={this.state.status}
                        />
                    </div>
                )}
            </div>
        );
    }
}

export default ProfileStatus;
