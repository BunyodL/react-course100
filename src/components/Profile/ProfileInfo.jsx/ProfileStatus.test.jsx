import { create } from 'react-test-renderer';
import ProfileStatus from './ProfileStatus';

describe('ProfileStatus component', () => {
  test('status from props should be in the local state status', () => {
    const component = create(<ProfileStatus status={'samuraiJS.com'} />);
    let instanse = component.getInstance();
    expect(instanse.props.status).toEqual(instanse.state.status);
  });

  test('span should be displayed after component creation',async  () => {
    const component = create(<ProfileStatus status={'samuraiJS.com'} />);
    let root = component.root;
    let span = root.findByType('span');
    expect(span).toBeDefined();
  });

  test("input shouldn't be displayed after component creation", () => {
    const component = create(<ProfileStatus status={'samuraiJS.com'} />);
    let root = component.root;
    expect(() => {
      root.findByType('input');
    }).toThrow();
  });

  test('span status should be correct', () => {
    const component = create(<ProfileStatus status={'samuraiJS.com'} />);
    let root = component.root;
    let instanse = component.getInstance();
    let span = root.findByType('span');
    expect(span.children[0]).toBe(instanse.state.status);
  });

  test('after double click input should be displayed instead of span', () => {
    const component = create(<ProfileStatus status={'samuraiJS.com'} />);
    let root = component.root;
    let span = root.findByType('span');
    span.props.onDoubleClick();
    let input = root.findByType('input');
    expect(input).toBeDefined();
  });

  test('input value should be the same with span status', () => {
    const component = create(<ProfileStatus status={'samuraiJS.com'} />);
    let root = component.root;
    let instanse = component.getInstance();
    let span = root.findByType('span');
    span.props.onDoubleClick();
    let input = root.findByType('input');
    expect(input.props.value).toBe(instanse.state.status);
  });

  test('callback should be called', () => {
    const mockCallBack = jest.fn();
    const component = create(<ProfileStatus status={'samuraiJS.com'} updateUserStatus={mockCallBack} />);
    let instanse = component.getInstance();
    instanse.deactivateEditMode();
    expect(mockCallBack.mock.calls.length).toBe(1);
  });
});
