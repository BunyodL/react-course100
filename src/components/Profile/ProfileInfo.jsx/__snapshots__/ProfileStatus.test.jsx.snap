// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pr component Matches the snapshot 1`] = `
<div
  className="profileStatus"
>
  <div>
    <span
      onDoubleClick={[Function]}
    >
      -----
    </span>
  </div>
</div>
`;
